store/v2/multi/snapshot.go

```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		return sdkerrors.Wrapf(err, "error while getting the snapshot versions at height %v", height)
+	}
+	if !versions.Exists(height) {
+		return sdkerrors.Wrapf(sdkerrors.ErrNotFound, "cannot find snapshot at height %v", height)
+	}
+
+	// get the saved snapshot at height
+	vs, err := rs.getView(int64(height))
+	if err != nil {
+		return sdkerrors.Wrap(err, fmt.Sprintf("error while get the version at height %d", height))
+	}
+
+	// schema keys
+	var sKeys []string
+	// sending the snapshot store schema
+	for sKey := range vs.schema {
+		if vs.schema[sKey] == storetypes.StoreTypePersistent {
+			sKeys = append(sKeys, sKey)
+		}
+	}
+
+	sort.Slice(sKeys, func(i, j int) bool {
+		return strings.Compare(sKeys[i], sKeys[j]) == -1
+	})
+
+	var storeByteKeys [][]byte
+	for _, sKey := range sKeys {
+		storeByteKeys = append(storeByteKeys, []byte(sKey))
+	}
+
+	err = protoWriter.WriteMsg(&snapshottypes.SnapshotItem{
+		Item: &snapshottypes.SnapshotItem_Schema{
+			Schema: &snapshottypes.SnapshotSchema{
+				Keys: storeByteKeys,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, sKey := range sKeys {
+		subStore, err := vs.getSubstore(sKey)
+		if err != nil {
+			return err
+		}
+
+		err = protoWriter.WriteMsg(&snapshottypes.SnapshotItem{
+			Item: &snapshottypes.SnapshotItem_Store{
+				Store: &snapshottypes.SnapshotStoreItem{
+					Name: sKey,
+				},
+			},
+		})
+		if err != nil {
+			return err
+		}
+
+		iter := subStore.Iterator(nil, nil)
+		for ; iter.Valid(); iter.Next() {
+			err = protoWriter.WriteMsg(&snapshottypes.SnapshotItem{
+				Item: &snapshottypes.SnapshotItem_KV{
+					KV: &snapshottypes.SnapshotKVItem{
+						Key:   iter.Key(),
+						Value: iter.Value(),
+					},
+				},
+			})
+			if err != nil {
+				return err
+			}
+		}
+
+		err = iter.Close()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Restore implements snapshottypes.Snapshotter.
+func (rs *Store) Restore(
+	height uint64, format uint32, protoReader protoio.Reader,
+) (snapshottypes.SnapshotItem, error) {
+	if err := snapshots.ValidRestoreHeight(format, height); err != nil {
+		return snapshottypes.SnapshotItem{}, err
+	}
+
+	versions, err := rs.stateDB.Versions()
+	if err != nil {
+		return snapshottypes.SnapshotItem{}, sdkerrors.Wrapf(err, "error while getting the snapshot versions at height %v", height)
+	}
+	if versions.Count() != 0 {
+		return snapshottypes.SnapshotItem{}, sdkerrors.Wrapf(sdkerrors.ErrLogic, "cannot restore snapshot for non empty store at height %v", height)
+	}
+
+	var subStore *substore
+	var storeSchemaReceived = false
+
+	for {
+		snapshotItem := &snapshottypes.SnapshotItem{}
+		err := protoReader.ReadMsg(snapshotItem)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return snapshottypes.SnapshotItem{}, sdkerrors.Wrap(err, "invalid protobuf message")
+		}
+
+		switch item := snapshotItem.Item.(type) {
+		case *snapshottypes.SnapshotItem_Schema:
+			if len(rs.schema) != 0 {
```

Comment:
We will actually need to change this behavior. StoreKeys are registered by the app at init, and exist as unique objects referenced only from modules that have access to them. (I recently corrected the v2 store to follow this pattern.) 

So, the store needs to have a schema containing all the stores received by the snapshot before the snapshot is received - this is the current behavior for the current MultiStore.

So here, instead of overwriting the store's schema, we'll just check that it matches the one received.

Reply:
okay